File: TicTacToeBoard.R
makeMove <- function(obj, move) { UseMethod("makeMove") }
isGameOver <- function(obj) { UseMethod("isGameOver") }
undoMove <- function(obj, move) { UseMethod("undoMove") }
getPosition <- function(obj) { UseMethod("getPosition") }

TicTacToeBoard <- function() {
	value <- list(whoToMove = "X", 
		      isGameOver = FALSE, 
		      winner = NA,
		      board = rep("-",9),
		      legalMoves = c("A1","A2","A3","B1","B2","B3","C1","C2","C3"),
		      moves <- list()
		     )
	attr(value, "class") <- "TicTacToeBoard"
	value
}

getPosition.TicTacToeBoard <- function(obj) {
	result <- paste0(obj$board,collapse="")
	result <- gsub("-","001",result,fixed = TRUE)
	result <- gsub("X","010",result,fixed = TRUE)
	result <- gsub("O","100",result,fixed = TRUE)
	result
}

undoMove.TicTacToeBoard <- function(obj, move) {
  row <- switch(substr(move,1,1), A = 0, B = 1, C = 2)
  col <- as.integer(substr(move,2,2))
  pos <- (3 * row + col)
  if (obj$board[[pos]] == "-") return(obj)
  obj$moves <- obj$moves[-length(obj$moves)]
  obj$board[[pos]] <- "-"

  if (obj$whoToMove == "X") {
    obj$whoToMove <-  "O"
  } else {
    obj$whoToMove <- "X"
  }
  obj$legalMoves <- append(obj$legalMoves, move)
  #obj <- isGameOver(obj)
  obj$isGameOver <- FALSE
  obj
}

isGameOver.TicTacToeBoard <- function(obj) {
	if (obj$isGameOver) { return(obj) }
	if (length(obj$legalMoves) == 0) {
		obj$isGameOver <- TRUE
		obj$winner <- "-"
	}
	for (j in 0:2) {
		x1 <- 0; o1 <- 0; x2 <- 0; o2 <- 0
		for (i in 0:2) {
			pos1 <- i+j*3+1
			pos2 <- j+i*3+1
			if (obj$board[[pos1]] == "X") { x1 <- x1 + 1 }
			if (obj$board[[pos1]] == "O") { o1 <- o1 + 1 }
			if (obj$board[[pos2]] == "X") { x2 <- x2 + 1 }
			if (obj$board[[pos2]] == "O") { o2 <- o2 + 1 }
		}
		if (x1== 3 || x2 == 3) {
			obj$isGameOver <- TRUE
			obj$winner <- "X"
		}
		if (o1== 3 || o2 == 3) {
			obj$isGameOver <- TRUE
			obj$winner <- "O"
		}
	}
	if (obj$board[[1]] == obj$board[[5]] && 
	    obj$board[[5]] == obj$board[[9]] &&
	    obj$board[[1]] == "X") {
		obj$isGameOver <- TRUE
		obj$winner <- "X"
	}
	if (obj$board[[1]] == obj$board[[5]] && 
	    obj$board[[5]] == obj$board[[9]] &&
	    obj$board[[1]] == "O") {
		obj$isGameOver <- TRUE
		obj$winner <- "O"
	}
	if (obj$board[[3]] == obj$board[[5]] && 
	    obj$board[[5]] == obj$board[[7]] &&
	    obj$board[[3]] == "X") {
		obj$isGameOver <- TRUE
		obj$winner <- "X"
	}
	if (obj$board[[3]] == obj$board[[5]] && 
	    obj$board[[5]] == obj$board[[7]] &&
	    obj$board[[3]] == "O") {
		obj$isGameOver <- TRUE
		obj$winner <- "O"
	}
	obj
}

print.TicTacToeBoard <- function(obj, full = FALSE) {
	cat(paste0("A ",obj$board[[1]],obj$board[[2]],obj$board[[3]],"\n"))
	cat(paste0("B ",obj$board[[4]],obj$board[[5]],obj$board[[6]],"\n"))
	cat(paste0("C ",obj$board[[7]],obj$board[[8]],obj$board[[9]],"\n"))
	cat(paste0("  123\n"))
	cat("\n")
	if (full) {
	  cat("Moves: ",obj$moves,"\n")
	  cat("Who to move: ",obj$whoToMove,"\n")
	  cat("isGameOver: ",obj$isGameOver,"\n")
	  cat("winner: ",obj$winner,"\n")
	  cat("legalMoves: ",obj$legalMoves,"\n")
	}
}

makeMove.TicTacToeBoard <- function(obj, move) {
	if (obj$isGameOver) return(obj)
	obj$moves <- append(obj$moves, move)
	row <- switch(substr(move,1,1), A = 0, B = 1, C = 2)
	col <- as.integer(substr(move,2,2))
	pos <- (3 * row + col)
	if (obj$board[[pos]] != "-") return(obj)
	if (obj$whoToMove == "X") {
		obj$board[[pos]] <- "X"
		obj$whoToMove <-  "O"
	} else {
		obj$board[[pos]] <- "O"
		obj$whoToMove <- "X"
	}
	obj$legalMoves <- obj$legalMoves[obj$legalMoves != move]
	obj <- isGameOver(obj)
	obj
}

playRandomGames <- function(num = 1) {
  replicate(num, {
    b <- TicTacToeBoard()
    while (b$isGameOver == FALSE) {
      aMove <- sample(b$legalMoves,1)
      b <- makeMove(b, aMove)
    }
    b$moves
  })
}

unittest <- function() {
  b <- TicTacToeBoard()
  print("EMPTY TEST:")
  print(b, full = T)
  cat("\n")
  b <- makeMove(b,"B3")
  b <- makeMove(b,"A3")
  print("B3 TEST:")
  print(b, full = T)
  cat("\n")
  b <- undoMove(b, "B3")
  print("UNDO TEST:")
  print(b, full = T)
  cat("\n")
}

File: code/C-sharp/TicTacToeBoard.cs
using System;
using System.Collections.Generic;

namespace TicTacToe
{	
	class TicTacToeBoard
	{
		private char _whoToMove;
		private bool is_game_over;
		private char winner;
		private string lastMove;
		private char[,] board;
		private List<string> emptyPlaces;
		private List<string> _playedMoves;

		public TicTacToeBoard()
		{
			_playedMoves=new List<string>();
			_whoToMove = 'X';
			is_game_over = false;
			winner = ' ';
			lastMove = "";
			board = new char[3,3];
			for (int row=0; row<3; row++)
				for (int col=0; col<3; col++)
					board[row,col] = '-';
			emptyPlaces=new List<string>();
			emptyPlaces.Add("A1");
			emptyPlaces.Add("A2");
			emptyPlaces.Add("A3");
			emptyPlaces.Add("B1");
			emptyPlaces.Add("B2");
			emptyPlaces.Add("B3");
			emptyPlaces.Add("C1");
			emptyPlaces.Add("C2");
			emptyPlaces.Add("C3");
		}

		public TicTacToeBoard(TicTacToeBoard t2) {
			_whoToMove=t2._whoToMove;
			is_game_over=t2.is_game_over;
			winner=t2.winner;
			lastMove=t2.lastMove;
			board=t2.board;
			emptyPlaces=t2.emptyPlaces;
			_playedMoves=t2._playedMoves;
		}
	
		public bool makeMove(string move) 
		{
			int row;
			int col;

			if (is_game_over) return false;
			move=move.ToUpper();
			row=char.Parse(move.Substring(0,1))-'A';
			col=char.Parse(move.Substring(1,1))-'1';
			if (board[row,col] != '-') return false;
			_playedMoves.Add(new string(move));
			board[row,col]=_whoToMove;
			if (_whoToMove == 'X') {
				_whoToMove='O';
			} else {
				_whoToMove='X';
			}
			emptyPlaces.Remove(move);
			lastMove=move;
			return isGameOver();
		}

		public override string ToString()
		{
			string result=new string("");
		
			result += "  123\n";
			result += "A "; 
			result += board[0,0];
			result += board[0,1];
			result += board[0,2];
			result += " A\n";
			result += "B "; 
			result += board[1,0];
			result += board[1,1];
			result += board[1,2];
			result += " B\n";
			result += "C "; 
			result += board[2,0];
			result += board[2,1];
			result += board[2,2];
			result += " C\n";
			result += "  123";
			return result;
		}

		public char whoToMove() { return _whoToMove; }

		public void undoLastMove()
		{
			int row, col;

			if (lastMove == " ") return;
			row=char.Parse(lastMove.Substring(0,1))-'A';
			col=char.Parse(lastMove.Substring(1,1))-'1';
			board[row,col]='-';
			if (_whoToMove == 'X') {
				_whoToMove='O';
			} else {
				_whoToMove='X';
			}
			emptyPlaces.Add(lastMove);
			_playedMoves.Remove(lastMove);
			if (_playedMoves.Count == 0) {
				lastMove=" ";
			} else {
				lastMove=_playedMoves[_playedMoves.Count-1];
			}
			is_game_over=false;
		}

		public List<string> legalMoves() 
		{
			List<string> result=new List<string>(emptyPlaces);
			return result;
		}

		public List<string> playedMoves()
		{
			return _playedMoves;
		}

		public int eval()
		{
			if (isGameOver()) {
				if (winner == 'X') return  100;
				if (winner == 'O') return -100;
			}
			return 0;
		}

		public bool isGameOver()
		{
			int Xcount;
			int Ocount;

			if (is_game_over) return true;
			if (emptyPlaces.Count == 0) {
				is_game_over=true;
				winner=' ';
				return true;
			}
			for (int col=0; col<3; col++) {
				Xcount=0;
				Ocount=0;
				for (int row=0; row<3; row++) {
					if (board[row,col] == 'X') Xcount++;
					if (board[row,col] == 'O') Ocount++;
				}
				if (Xcount == 3) {
					is_game_over=true;
					winner = 'X';
					return true;
				}
				if (Ocount == 3) {
					is_game_over=true;
					winner = 'O';
					return true;
				}
			}
			for (int row=0; row<3; row++) {
				Xcount=0;
				Ocount=0;
				for (int col=0; col<3; col++) {
					if (board[row,col] == 'X') Xcount++;
					if (board[row,col] == 'O') Ocount++;
				}
				if (Xcount == 3) {
					is_game_over=true;
					winner = 'X';
					return true;
				}
				if (Ocount == 3) {
					is_game_over=true;
					winner = 'O';
					return true;
				}
			}
			if (board[0,0] == board[1,1] &&
			    board[0,0] == board[2,2] &&
			    board[0,0] != '-') {
			  is_game_over=true;
			  winner=board[0,0];
			  return true;
			}
			if (board[0,2] == board[1,1] &&
			    board[0,2] == board[2,0] &&
			    board[0,2] != '-') {
			  is_game_over=true;
			  winner=board[0,2];
			  return true;
			}
			return false;
		}
	}

}
